### PR TITLE
Added Shiro security

### PR DESCRIPTION
**Pull request by hayssams**
Added shiro security. HTTP calls and websockets communications are both protected.

### CODE REVIEW

**Comment by Leemoonsoo** [comment]
Hi @hayssams, Thanks for contribution. Let me test this branch.

**Comment by hayssams** [comment]
Hi @Leemoonsoo, did you get a chance to test this PR ?

**Comment by Leemoonsoo** [comment]
Sorry for late responding and thanks again for contribution.

In my understanding,  this branch trying to provides 1) security of rest/websocket api. 2) multi-tenant environment for notebook save/load.

Because of this PR brings wide range of changes, we have couple of things to discuss.
  - providing login window
  - displaying current user
  - how to merge changes https://issues.apache.org/jira/browse/ZEPPELIN-26
  - migration script for notebook directory structure change.
  - notebook multi tenant model (per user / shares all, due to any user actually can read any notebook from filesystem api on Spark/Pyspark interpreter)

I'm feeling like if it is possible to separate 1) security of rest/websocket  2) multi-tenancy of notebook into different issue, it'll be much faster to review and test.

What do you think?

**Comment by hayssams** [comment]
You're right, this PR addresses 1) & 2).

Regarding the points you raised
- The login window is useful if we want to support form auth in addition to the basic auth this PR provides. This may be added in another PR easily with almost no impact to the current code.

- Right, we need to display the current user. Do you want me to add it at the top of the note with a logout button ?
- Zeppelin-26 : Did not look at it yet. 
- Migration script : that should be easy. to just move existing notes to the anonymous subdirectory with no impact on the existing code
- Multitenant : Would starting the "interpreter.sh" script with a different identity solve the problem ?

Actually feature 2) involves a very small amount of code (load/save notes in user directory).

I am personally interested by feature 2). let me know if you want me to complete the other points you raised first or if you prefer to stick to feature 1). I'll respect your choice.

**Comment by hayssams** [comment]
Hi Anthony 

1. You're right using cookies for the HTTP API would be a much better choice.

2. Regarding storm path, The comments you find in the shiro.ini file come from the default shiro.ini setup; Wen can safely ignore/remove them

3. The login page is not required for shiro to work. We can use basic auth by uncommenting line 17 in the shiro.ini file (/** = authcBasic). The browser will popup a login dialog without any coding required.

4. By default auth is not activated. To activate it, simply comment line 16 and uncomment line 17 in shiro.ini. The file shiro.ini need to be accessible through the classpath.

**Comment by hayssams** [comment]
I missed your last question about session handling.
I do not use sessions since the credentials are passed with each call.

**Comment by hayssams** [comment]
Hi
I updated the PR to include the following
1. Merged with master (https://issues.apache.org/jira/browse/ZEPPELIN-26)
2. Credentials are included as HTTP headers to access the REST API
3. No more stormpath comments in the shiro.ini file
4. Current user is now displayed on the main screen

**Comment by hayssams** [comment]
Hi
No credentials needed to access the REST API. Principal comes now from shiro context

**Comment by RPCMoritz** [comment]
Hi, I'd like to know whether this PR will be followed up? Integrated authentication would go along way towards making Zeppelin enterprise capable without using the current crutches that are required.

**Comment by hayssams** [comment]
Hi, I'll merge this WE nad let you know as soon as it's done.

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java [hunk 1]
```diff
@@ -73,10 +81,18 @@ public Response bind(@PathParam("noteId") String noteId, String req) throws IOEx
   @GET
   @Path("interpreter/bind/{noteId}")
   public Response bind(@PathParam("noteId") String noteId) {
-    List<InterpreterSettingListForNoteBind> settingList
-      = new LinkedList<InterpreterSettingListForNoteBind>();
+    Object oprincipal = SecurityUtils.getSubject().getPrincipal();
```
Comment (djoelz) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: This should be its own method as its reused.

**Comment by djoelz** [comment]
I feel that this needs far more test coverage with all the new public methods it introduces. Just changing the existing tests is not enough. Please add unit tests to the methods added/changed in NotebookRestApi
SecurityRestApi
ZeppelinServer
NotebookServer
TicketContainer
Notebook
VFSNotebookRepo

Also we should think about adding security with interfaces and have this be an implementation of those interfaces. Thoughts?

**Comment by hayssams** [comment]
Hello
Just came back from vacation. Give me a couple of weeks to build the
required tests.

Cordialement.

*Hayssam Saleh, Architecture & Stratégie S.I.
*+33 6 85 71 13 95 | LinkedIn Profile
<http://fr.linkedin.com/pub/hayssam-saleh/0/144/bb8> | ebiznext :)
<http://www.ebiznext.com> | Skype ID <http://hayssam.saleh2>


On Tue, Aug 18, 2015 at 8:40 PM, djoelz <notifications@github.com> wrote:

> I feel that this needs far more test coverage with all the new public
> methods it introduces. Just changing the existing tests is not enough.
> Please add unit tests to the methods added/changed in NotebookRestApi
> SecurityRestApi
> ZeppelinServer
> NotebookServer
> TicketContainer
> Notebook
> VFSNotebookRepo
>
> Also we should think about adding security with interfaces and have this
> be an implementation of those interfaces. Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/53#issuecomment-132313144>
> .
>

-- 
*EBIZNEXT <http://www.ebiznext.com>  Java / .NET **/ DevOps / BigData / UX 
/ Agile*
23 rue Vernet 75008 Paris 
<https://maps.google.com/?q=23%20rue%20Vernet%2075008%20Paris>

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java (hunk 1 above)
Comment (hayssams) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: OK. will do it shortly

Cordialement.

*Hayssam Saleh, Architecture & Stratégie S.I.
*+33 6 85 71 13 95 | LinkedIn Profile
<http://fr.linkedin.com/pub/hayssam-saleh/0/144/bb8> | ebiznext :)
<http://www.ebiznext.com> | Skype ID <http://hayssam.saleh2>


On Tue, Aug 18, 2015 at 8:30 PM, djoelz <notifications@github.com> wrote:

> In
> zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java
> <https://github.com/apache/incubator-zeppelin/pull/53#discussion_r37335212>
> :
>
> > @@ -73,10 +81,18 @@ public Response bind(@PathParam("noteId") String noteId, String req) throws IOEx
> >    @GET
> >    @Path("interpreter/bind/{noteId}")
> >    public Response bind(@PathParam("noteId") String noteId) {
> > -    List<InterpreterSettingListForNoteBind> settingList
> > -      = new LinkedList<InterpreterSettingListForNoteBind>();
> > +    Object oprincipal = SecurityUtils.getSubject().getPrincipal();
>
> This should be its own method as its reused.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/53/files#r37335212>.
>

-- 
*EBIZNEXT <http://www.ebiznext.com>  Java / .NET **/ DevOps / BigData / UX 
/ Agile*
23 rue Vernet 75008 Paris 
<https://maps.google.com/?q=23%20rue%20Vernet%2075008%20Paris>

**Comment by MartinWeindel** [comment]
@hayssams:
There is a small bug, if you from a notebook to the home page by clicking on the top left Zeppelin link.
In this case the user is set to anonymous.

I suggest the patch below to fix this problem.

```
diff --git a/zeppelin-web/src/app/home/home.controller.js b/zeppelin-web/src/app/home/home.controller.js
index 1a20219..77f8185 100644
--- a/zeppelin-web/src/app/home/home.controller.js
+++ b/zeppelin-web/src/app/home/home.controller.js
@@ -14,11 +14,13 @@
 'use strict';
 
 angular.module('zeppelinWebApp').controller('HomeCtrl', function($scope, notebookListDataFactory, websocketMsgSrv, $rootScope, arrayOrderingSrv) {
-  $rootScope.ticket = {
-    'principal':'anonymous',
-    'ticket':'anonymous'
-  };
-
+  if (!$rootScope.ticket) {
+	  $rootScope.ticket = {
+	    'principal':'anonymous',
+	    'ticket':'anonymous'
+	  };
+  }
+  
   var vm = this;
   vm.notes = notebookListDataFactory;
   vm.websocketMsgSrv = websocketMsgSrv;
diff --git a/zeppelin-web/src/components/navbar/navbar.controller.js b/zeppelin-web/src/components/navbar/navbar.controller.js
index 73be946..9be6b9b 100644
--- a/zeppelin-web/src/components/navbar/navbar.controller.js
+++ b/zeppelin-web/src/components/navbar/navbar.controller.js
@@ -16,10 +16,12 @@
 'use strict';
 
 angular.module('zeppelinWebApp').controller('NavCtrl', function($scope, $rootScope, $routeParams, notebookListDataFactory, websocketMsgSrv, arrayOrderingSrv, $http) {
-  $rootScope.ticket = {
-    'principal':'anonymous',
-    'ticket':'anonymous'
-  };
+  if (!$rootScope.ticket) {
+	  $rootScope.ticket = {
+			    'principal':'anonymous',
+			    'ticket':'anonymous'
+			  };
+  }
   /** Current list of notes (ids) */
 
   var vm = this;
```

**Comment by hayssams** [comment]
LGTM.

Just pushed the update;
Thanks
Hayssam

>

-- 
*EBIZNEXT <http://www.ebiznext.com>  Java / .NET **/ DevOps / BigData / UX 
/ Agile*
23 rue Vernet 75008 Paris 
<https://maps.google.com/?q=23%20rue%20Vernet%2075008%20Paris>

**Comment by lazaromedina** [comment]
Hi @hayssams , 
i think there is another small bug. If a user different from 'anonymous' is authenticated, and clicking on one of his notes and then refreshing from webexplorer, the main board with paragraphs is lost, but upper panel shows that user is still authenticated. It seems that refreshing sends again to server information about anonymous and this one has nothing about noteId of user authenticated previously.
Hope i can help, thanks for the pull.

**Comment by doanduyhai** [comment]
+ 100 for this feature. Security is missing right now and that prevents people for deploying Zeppelin in production.

**Comment by jeffsteinmetz** [comment]
Does this also handle securing the websocket?  Could Shiro be extended to support single sign on tokens.  For example - a future option where authentication is performed outside of zeppelin/shiro, and a java web token in the url.

**Comment by hayssams** [comment]
Hello Jeff
Yes websockets are secured. To get added to the user/websocket map, the client must provide the valid ticket. See code below:
      LOG.info("RECEIVE OP << " + messagereceived.op);
      LOG.info("RECEIVE PRINCIPAL << " + messagereceived.principal);
      LOG.info("RECEIVE TICKET << " + messagereceived.ticket);
      String ticket = TicketContainer.instance.getTicket(messagereceived.principal);
      if (ticket != null && !ticket.equals(messagereceived.ticket))
        throw new Exception("Invalid ticket " + messagereceived.ticket + " != " + ticket);

      addConnectionToUserSocketMap(conn, messagereceived);


Regarding Shiro / SSO, you'll find many integrations around. And yes it may be extended without code change to the Zeppelin code base.

**Comment by galleon** [comment]
Hayssam, what still needs to be done on this PR so that it can be merged?

**Comment by jeffsteinmetz** [comment]
Can you add documentation on how to use this?  perhaps a new SECURITY-README.md and a link to it from the main readme?  or maybe something new in the dev docs subdirectory?

**Comment by hayssams** [comment]
Hi Galleon

I have been asked to add a few more unit tests by djoelz above. I am halfway in finishing it.

**Comment by galleon** [comment]
Hi, thanks for your answer. Let me know if I can help. Cheers. Guillaume

**Comment by msciab** [comment]
Hi, I just built your branch and I also enabled authentication editing shiro.ini. Now I get the browser asking for the password, but I can just cancel  the request and I am able to do everything, including executing scala code. Maybe I am missing something about how this patch is supposed to work...

**Comment by hayssams** [comment]
By cancelling you get access as an anonymous user aka public notes only.

**Comment by msciab** [comment]
Yes but cancelling I am still able to create a public note, and in a public note I can still write whatsoever code. It may protect the notes but it does not protect the system from being abused... Is it not possible to block access to not-authenticated users?

**Comment by hayssams** [comment]
Hi
I just added a boolean property (zeppelin.anonymous.allowed) at the end of the file conf/zeppelin-site.xml that when to false will block not-auhtenticated users.

**Comment by sejunra** [comment]
wondering what the status is on this. this is an awesome feature being awaited by many. thanks.

**Comment by elbamos** [comment]
I think we should have an organized plan to review and address this PR.  

I propose that the PR be merged into a new branch of Zeppelin, where the community as a whole will have an easier time reviewing and making PRs on the branch.

Then I propose that we review the aspects of this separately.  By reviewing/approving/voting-on each aspect in sequence, we will have a clear idea of where the PR is in process, and that will make it easier to move PR forward to final merge (if that's what we want to do).  The aspects I see are:
* Feature set -- is this the kind of security interface we want to put in Zeppelin?
* Architecture -- Is this (shiro) how we want to implement security?
* Does it work? 
* Do we have adequate tests to be able to support and maintain it?
* Is it adequately documented that people will be able to use it?

The reason I propose a more formal process is that this PR has been sitting here for 8 months now.  With that, and with some 82 (!!!) commits, a more structured approach seems like the best way to resolve this PR (one way or the other).  

(In full disclosure -- My only view about this is that since its security-related, it needs to have *mounds* of tests to make sure the security doesn't get compromised by later commits.)

**Comment by jongyoul** [comment]
In addition to @elbamos's comments, I hope that the community would understand that it's too hard to review a kind of this PR which has changed widely because committers may not know features well. I, however, also know some features take extensive changes. In this case, I propose that a contributor makes a feature Jira issue that includes several sub-tasks, and makes a - relatively - small PR for every sub-task. It would help review, test and merge the PR easily.

**Comment by rconline** [comment]
@jongyoul @elbamos @hayssams this change is an important one, and we should try to get this in. My two cents on how we could divide the whole pull request into:

- inclusion of libraries and dependencies
- Addition of security filter, changes in Angular Object, configurable - turned off by default. With a basic set of test-cases, which ensure the changes are non-breaking. 
- Front-end changes, which include -display of username, accepting header parameters of principal, ticket. 
- Subsequently, addition of more test cases, before we call it done. Till such time we keep the feature as default off. 
(Each of the above steps will have user guides/documentation). 

If the committers can provide some bandwidth for review, I'm willing to make the above changes + documentation. 

As side notes:
- We will need to include the notion of proxy-users at sometime, which enable command execution of secure spark/hdfs clusters
- Subsequently, there will have to be a mechanism of how to provide default access to the cluster users, either by way of import of AD/LDAP users, through some auth_to_local kind of rules. 

I'm of the opinion that Shiro is a good strategy, larger projects such as https://knox.apache.org/ use Shiro, for HDFS Api gateway security.

**Comment by hayssams** [comment]
@elbamos @jongyoul @rconline 
Let me write down the whole set of changes I've made and how they may be split in my view in smaller commits.
There are also different strategies in using shiro. I would like to submit them and take into account the one you advise to use.

**Comment by jongyoul** [comment]
@hayssams Sure! You can split this into several commits because this PR is your contribution and we will thank you so much of that. If you want to do this, would you mind creating Jira issue or linking a related issue that is already created?

**Comment by jeffsteinmetz** [comment]
Is there an example of how shiro would be pluggable with other forms of
authentication and authorization?
Such as Java Web tokens (JWT), which require storing a secret.  And the JWT
contains authorization roles.

**Comment by rconline** [comment]
@hayssams please do that, it will be great. I can help with Junits, Docs and separation, should you need it, will commit into your branches, which can then be submitted as PRs.

**Comment by hayssams** [comment]
I have just added a short  [security readme](https://github.com/ebiznext/incubator-zeppelin/blob/3b276057a22c5569ea6f6cf1c5851b4aa5277e2f/Security-README.md) that may allow us to choose the best strategy in adding Shiro. Basically the two questions are :
1. Is the way we handle web socket security a good pattern
2. Would we rely on the ThreadLocal pattern or explicitly pass the username in function calls or should we use a different approach ?

Once we decide how to move on, I think that we will be able to focus on the implementation details and how this PR should be split.

Waiting to hear from you.

**Comment by elbamos** [comment]
@hayssams Can you provide a very step-by-step intro readme?  Default username/pass, what the authentication backend is, how users/groups are added, how to make notebooks belong to one user or another...  That very basic stuff?

**Comment by hayssams** [comment]
@elbamos @jongyoul @rconline 
I created a new PR (https://github.com/apache/incubator-zeppelin/pull/586) which implements HTTP et Websocket security but does not implement note ownership. This means that a user must authenticate to access Zeppelin but once authenticated, he will have access to all notes. That's what @Leemoonsoo wanted to do first.

For easier review, this new PR (#586) contains only two commits. The first one implements HTTP Auth and the second one Websocket authentication . 

Once  PR # 586 is merged / validated, I suggest that we move forward with note ownership and permissions.

**Comment by jeffsteinmetz** [comment]
One comment:  why is Stormpath (a commercial auth service unrelated to Zeppelin) mentioned in the comments for this PR?
See:
https://github.com/apache/incubator-zeppelin/pull/586/files#diff-9b7b90bbe0069ba81385d11637dcd4e4R27


From:  Hayssam Saleh
Reply-To:  apache/incubator-zeppelin
Date:  Thursday, December 31, 2015 at 10:06 AM
To:  apache/incubator-zeppelin
Cc:  Jeff Steinmetz
Subject:  Re: [incubator-zeppelin] Added Shiro security (#53)


@elbamos @jongyoul @rconline 
I created a new PR (#586) which implements HTTP et Websocket security but does not implement note ownership. This means that a user must authenticate to access Zeppelin but once authenticated, he will have access to all notes. That's what @Leemoonsoo wanted to do first.

For easier review, this new PR (#586) contains only two commits. The first one implements HTTP Auth and the second one Websocket authentication . 

Once  PR # 586 is merged / validated, I suggest that we move forward with note ownership and permissions.

**Comment by hayssams** [comment]
You are right that should be removed.
I took it from the shiro sample.
I am removing the whole section right away.
Cordialement.
Hayssam Saleh, CTO
+33 6 85 71 13 95 <> | LinkedIn Profile <http://fr.linkedin.com/pub/hayssam-saleh/0/144/bb8> | ebiznext :) <http://www.ebiznext.com/> | Skype ID <http://hayssam.saleh2/>



> On 31 Dec 2015, at 20:27, Jeff Steinmetz <notifications@github.com> wrote:
> 
> One comment: why is Stormpath (a commercial auth service unrelated to Zeppelin) mentioned in the comments for this PR?
> See:
> https://github.com/apache/incubator-zeppelin/pull/586/files#diff-9b7b90bbe0069ba81385d11637dcd4e4R27
> 
> 
> From: Hayssam Saleh
> Reply-To: apache/incubator-zeppelin
> Date: Thursday, December 31, 2015 at 10:06 AM
> To: apache/incubator-zeppelin
> Cc: Jeff Steinmetz
> Subject: Re: [incubator-zeppelin] Added Shiro security (#53)
> 
> 
> @elbamos @jongyoul @rconline 
> I created a new PR (#586) which implements HTTP et Websocket security but does not implement note ownership. This means that a user must authenticate to access Zeppelin but once authenticated, he will have access to all notes. That's what @Leemoonsoo wanted to do first.
> 
> For easier review, this new PR (#586) contains only two commits. The first one implements HTTP Auth and the second one Websocket authentication . 
> 
> Once PR # 586 is merged / validated, I suggest that we move forward with note ownership and permissions.
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> 
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/apache/incubator-zeppelin/pull/53#issuecomment-168232632>.
> 


-- 
*EBIZNEXT <http://www.ebiznext.com>  Java / .NET **/ DevOps / BigData / UX 
/ Agile*
23 rue Vernet 75008 Paris 
<https://maps.google.com/?q=23%20rue%20Vernet%2075008%20Paris>

**Comment by hayssams** [comment]
@jongyoul 
I just created 2 JIRA issues.
[ZEPPELIN-548](https://issues.apache.org/jira/browse/ZEPPELIN-548) for Zeppelin Authentication
[ZEPPELIN-549](https://issues.apache.org/jira/browse/ZEPPELIN-549) for Zeppelin Authorisation

**Comment by hayssams** [comment]
I am closing this PR and suggest that we move the discussion to #586

**Comment by guptabandhu25** [comment]
Hi Hayssams , the authentication letter is applied but in terms of exception as  i am not getting any window for putting my username and password . I am in the customer network so it is not asking for any authentication .

Showing the message 

ERROR [2016-06-02 14:27:22,265] ({qtp513169028-34} NotebookServer.java[onMessage]:179) - Can't handle message
java.lang.Exception: Anonymous access not allowed
        at org.apache.zeppelin.socket.NotebookServer.onMessage(NotebookServer.java:110)
        at org.apache.zeppelin.socket.NotebookSocket.onMessage(NotebookSocket.java:56)
        at org.eclipse.jetty.websocket.WebSocketConnectionRFC6455$WSFrameHandler.onFrame(WebSocketConnectionRFC6455.java:835)
        at org.eclipse.jetty.websocket.WebSocketParserRFC6455.parseNext(WebSocketParserRFC6455.java:349)
        at org.eclipse.jetty.websocket.WebSocketConnectionRFC6455.handle(WebSocketConnectionRFC6455.java:225)
        at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:667)
        at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:52)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:608)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:543)
        at java.lang.Thread.run(Thread.java:745)

**Pull request closed without merging**